File: plugins/home-assistant-dev/scripts/validate-strings.py
#!/usr/bin/env python3
"""Validate Home Assistant integration strings.json.

Checks for:
- Sync between config_flow.py steps and strings.json
- Required error keys
- Required abort keys
- Proper structure
"""
from __future__ import annotations

import ast
import json
import re
import sys
from pathlib import Path
from typing import Any

# Common error keys that should be present
COMMON_ERRORS = {"cannot_connect", "invalid_auth", "unknown"}

# Common abort reasons
COMMON_ABORTS = {"already_configured"}


class ValidationError:
    """Represents a validation error."""

    def __init__(self, field: str, message: str, severity: str = "error") -> None:
        self.field = field
        self.message = message
        self.severity = severity

    def __str__(self) -> str:
        return f"[{self.severity.upper()}] {self.field}: {self.message}"


def extract_flow_steps(config_flow_path: Path) -> tuple[set[str], set[str], set[str]]:
    """Extract step names, error keys, and abort reasons from config_flow.py.

    Returns:
        Tuple of (step_names, error_keys, abort_reasons)
    """
    steps: set[str] = set()
    errors: set[str] = set()
    aborts: set[str] = set()

    if not config_flow_path.exists():
        return steps, errors, aborts

    content = config_flow_path.read_text()

    # Find async_step_* methods
    step_pattern = re.compile(r"async def async_step_(\w+)\s*\(")
    for match in step_pattern.finditer(content):
        steps.add(match.group(1))

    # Find error assignments: errors["base"] = "cannot_connect"
    error_pattern = re.compile(r'errors\s*\[\s*["\'](\w+)["\']\s*\]\s*=\s*["\'](\w+)["\']')
    for match in error_pattern.finditer(content):
        errors.add(match.group(2))

    # Find abort calls: self.async_abort(reason="already_configured")
    abort_pattern = re.compile(r'async_abort\s*\(\s*reason\s*=\s*["\'](\w+)["\']')
    for match in abort_pattern.finditer(content):
        aborts.add(match.group(1))

    # Also check for _abort_if_unique_id_configured which uses "already_configured"
    if "_abort_if_unique_id_configured" in content:
        aborts.add("already_configured")

    return steps, errors, aborts


def validate_strings(
    strings_path: Path, config_flow_path: Path | None = None
) -> list[ValidationError]:
    """Validate strings.json file.

    Args:
        strings_path: Path to strings.json
        config_flow_path: Optional path to config_flow.py for sync checking

    Returns:
        List of validation errors
    """
    errors: list[ValidationError] = []

    # Check file exists
    if not strings_path.exists():
        errors.append(ValidationError("file", "strings.json not found"))
        return errors

    # Parse JSON
    try:
        with open(strings_path) as f:
            strings: dict[str, Any] = json.load(f)
    except json.JSONDecodeError as e:
        errors.append(ValidationError("json", f"Invalid JSON: {e}"))
        return errors

    # Check basic structure
    if "config" not in strings:
        errors.append(ValidationError("config", "Missing 'config' section"))
        return errors

    config = strings.get("config", {})
    if not isinstance(config, dict):
        errors.append(ValidationError("config", "Must be an object"))
        return errors

    # Check for step section
    if "step" not in config:
        errors.append(ValidationError("config.step", "Missing 'step' section"))

    # Check for error section
    if "error" not in config:
        errors.append(
            ValidationError("config.error", "Missing 'error' section", "warning")
        )

    # Check for abort section
    if "abort" not in config:
        errors.append(
            ValidationError("config.abort", "Missing 'abort' section", "warning")
        )

    # If we have config_flow.py, check sync
    if config_flow_path and config_flow_path.exists():
        flow_steps, flow_errors, flow_aborts = extract_flow_steps(config_flow_path)
        step_section = config.get("step", {})
        error_section = config.get("error", {})
        abort_section = config.get("abort", {})
        string_steps = set(step_section.keys()) if isinstance(step_section, dict) else set()
        string_errors = set(error_section.keys()) if isinstance(error_section, dict) else set()
        string_aborts = set(abort_section.keys()) if isinstance(abort_section, dict) else set()

        # Check for missing step strings
        # Exclude common internal steps that don't need strings
        internal_steps = {"reauth", "reauth_confirm", "reconfigure", "reconfigure_confirm"}
        for step in flow_steps:
            if step not in string_steps and step not in internal_steps:
                # Check if it might be using a different naming
                if step not in internal_steps:
                    errors.append(
                        ValidationError(
                            f"config.step.{step}",
                            f"Step '{step}' in config_flow.py but not in strings.json",
                            "warning",
                        )
                    )

        # Check for orphaned step strings
        for step in string_steps:
            if step not in flow_steps:
                errors.append(
                    ValidationError(
                        f"config.step.{step}",
                        f"Step '{step}' in strings.json but not in config_flow.py",
                        "warning",
                    )
                )

        # Check for missing error strings
        for error in flow_errors:
            if error not in string_errors:
                errors.append(
                    ValidationError(
                        f"config.error.{error}",
                        f"Error '{error}' used in config_flow.py but not in strings.json",
                    )
                )

        # Check for missing abort strings
        for abort in flow_aborts:
            if abort not in string_aborts:
                errors.append(
                    ValidationError(
                        f"config.abort.{abort}",
                        f"Abort reason '{abort}' used in config_flow.py but not in strings.json",
                    )
                )

    # Validate step structure
    steps = config.get("step", {})
    if not isinstance(steps, dict):
        errors.append(ValidationError("config.step", "Must be an object"))
        steps = {}
    for step_name, step_data in steps.items():
        if not isinstance(step_data, dict):
            errors.append(
                ValidationError(f"config.step.{step_name}", "Step must be an object")
            )
            continue

        # Check for data_description (Bronze requirement)
        if "data" in step_data and "data_description" not in step_data:
            errors.append(
                ValidationError(
                    f"config.step.{step_name}.data_description",
                    "Missing data_description (required for IQS Bronze)",
                    "warning",
                )
            )

        # Validate data and data_description keys match
        if "data" in step_data and "data_description" in step_data:
            data = step_data["data"]
            data_description = step_data["data_description"]
            if not isinstance(data, dict):
                errors.append(
                    ValidationError(f"config.step.{step_name}.data", "Must be an object")
                )
            elif not isinstance(data_description, dict):
                errors.append(
                    ValidationError(
                        f"config.step.{step_name}.data_description", "Must be an object"
                    )
                )
            else:
                data_keys = set(data.keys())
                desc_keys = set(data_description.keys())

                missing_desc = data_keys - desc_keys
                extra_desc = desc_keys - data_keys

                for key in missing_desc:
                    errors.append(
                        ValidationError(
                            f"config.step.{step_name}.data_description.{key}",
                            f"Missing description for field '{key}'",
                            "warning",
                        )
                    )

                for key in extra_desc:
                    errors.append(
                        ValidationError(
                            f"config.step.{step_name}.data_description.{key}",
                            f"Description for non-existent field '{key}'",
                            "warning",
                        )
                    )

    # Check options flow
    if "options" in strings:
        options = strings["options"]
        if not isinstance(options, dict):
            errors.append(ValidationError("options", "Must be an object"))
        elif "step" not in options:
            errors.append(
                ValidationError("options.step", "Missing 'step' section in options")
            )

    return errors


def main() -> int:
    """Main entry point."""
    if len(sys.argv) < 2:
        # Try to find strings.json
        search_paths = [
            Path("strings.json"),
            Path("custom_components") / "*" / "strings.json",
        ]
        strings_path = None
        for pattern in search_paths:
            matches = list(Path(".").glob(str(pattern)))
            if matches:
                strings_path = matches[0]
                break

        if strings_path is None:
            print("Usage: validate-strings.py <path/to/strings.json>")
            return 1
    else:
        strings_path = Path(sys.argv[-1])

    # Look for config_flow.py in same directory
    config_flow_path = strings_path.parent / "config_flow.py"

    errors = validate_strings(strings_path, config_flow_path)

    if not errors:
        return 0

    error_count = sum(1 for e in errors if e.severity == "error")
    warning_count = sum(1 for e in errors if e.severity == "warning")

    for error in errors:
        print(error)

    print()
    if error_count > 0:
        print(f"❌ {error_count} error(s), {warning_count} warning(s)")
        return 1
    else:
        print(f"⚠️  {warning_count} warning(s)")
        return 0


if __name__ == "__main__":
    sys.exit(main())

File: plugins/home-assistant-dev/mcp-server/src/tools/validate-strings.ts
/**
 * validate_strings tool - Validate strings.json and sync with config_flow.py
 */

import { readFile } from "fs/promises";
import { existsSync } from "fs";
import { dirname, join } from "path";
import type { ValidateStringsInput, ValidateStringsOutput } from "../types.js";

export async function handleValidateStrings(
  input: ValidateStringsInput
): Promise<ValidateStringsOutput> {
  // Return a structured invalid result for file/JSON errors (matching the Python twin
  // and validate-manifest), rather than throwing.
  const failure = (message: string): ValidateStringsOutput => ({
    valid: false,
    errors: [message],
    missing_steps: [],
    orphaned_steps: [],
    missing_errors: [],
    missing_data_descriptions: [],
  });

  // Check file exists
  if (!existsSync(input.path)) {
    return failure(`strings.json not found: ${input.path}`);
  }

  // Parse strings.json
  let strings: Record<string, unknown>;
  try {
    const content = await readFile(input.path, "utf-8");
    strings = JSON.parse(content);
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    return failure(`Invalid JSON in strings.json: ${message}`);
  }

  // Extract steps, errors, aborts from strings.json
  const config = strings.config as Record<string, unknown> | undefined;
  const stringSteps = new Set(
    Object.keys((config?.step as Record<string, unknown>) || {})
  );
  const stringErrors = new Set(
    Object.keys((config?.error as Record<string, unknown>) || {})
  );
  const stringAborts = new Set(
    Object.keys((config?.abort as Record<string, unknown>) || {})
  );

  // Try to load config_flow.py
  const configFlowPath = join(dirname(input.path), "config_flow.py");
  const flowSteps = new Set<string>();
  const flowErrors = new Set<string>();
  const flowAborts = new Set<string>();

  if (existsSync(configFlowPath)) {
    const flowContent = await readFile(configFlowPath, "utf-8");

    // Extract async_step_* methods
    const stepMatches = flowContent.matchAll(/async def async_step_(\w+)\s*\(/g);
    for (const match of stepMatches) {
      flowSteps.add(match[1]);
    }

    // Extract error keys: errors["base"] = "cannot_connect"
    // LIMITATION (literal-only): this matches only string-LITERAL assignments. Common
    // real-world forms — errors["base"] = err_key (variable) or dict comprehensions —
    // are silently missed, so those error keys cannot be cross-checked against strings.json
    // and may produce a false 'valid'. Broadening this safely requires real Python parsing.
    const errorMatches = flowContent.matchAll(
      /errors\s*\[\s*["'](\w+)["']\s*\]\s*=\s*["'](\w+)["']/g
    );
    for (const match of errorMatches) {
      flowErrors.add(match[2]);
    }

    // Extract abort reasons: both keyword (async_abort(reason="x")) and positional
    // (async_abort("x") / self.async_abort('x')) forms. Same literal-only caveat as
    // errors above — a variable-supplied reason is not captured.
    const abortMatches = flowContent.matchAll(
      /async_abort\s*\(\s*(?:reason\s*=\s*)?["'](\w+)["']/g
    );
    for (const match of abortMatches) {
      flowAborts.add(match[1]);
    }

    // Check for _abort_if_unique_id_configured
    if (flowContent.includes("_abort_if_unique_id_configured")) {
      flowAborts.add("already_configured");
    }
  }

  // Compare and find issues
  const internalSteps = new Set(["reauth", "reauth_confirm", "reconfigure", "reconfigure_confirm"]);

  const missingSteps: string[] = [];
  const orphanedSteps: string[] = [];
  const missingErrors: string[] = [];
  const missingDataDescriptions: string[] = [];

  // Check for missing step strings
  for (const step of flowSteps) {
    if (!stringSteps.has(step) && !internalSteps.has(step)) {
      missingSteps.push(step);
    }
  }

  // Check for orphaned step strings
  for (const step of stringSteps) {
    if (!flowSteps.has(step) && flowSteps.size > 0) {
      orphanedSteps.push(step);
    }
  }

  // Check for missing error strings
  for (const error of flowErrors) {
    if (!stringErrors.has(error)) {
      missingErrors.push(error);
    }
  }

  // Check for missing data_description (IQS Bronze requirement)
  // A hand-edited but still-parseable strings.json can place a non-object under a step
  // (e.g. "user": null or a string). Object.keys(null) throws, so guard the step entry
  // and each sub-field with object checks: a malformed step is reported as a structured
  // error instead of crashing the tool (PURPOSE 3: no crash on malformed input).
  const malformedSteps: string[] = [];
  const steps = (config?.step as Record<string, unknown>) || {};
  const isObject = (v: unknown): v is Record<string, unknown> =>
    typeof v === "object" && v !== null && !Array.isArray(v);
  for (const [stepName, rawStepData] of Object.entries(steps)) {
    if (!isObject(rawStepData)) {
      malformedSteps.push(`config.step.${stepName} must be an object`);
      continue;
    }
    const stepData = rawStepData;
    const data = stepData.data;
    const dataDescription = stepData.data_description;
    if (isObject(data) && !isObject(dataDescription)) {
      missingDataDescriptions.push(stepName);
    } else if (isObject(data) && isObject(dataDescription)) {
      // Check if all data keys have descriptions
      const dataKeys = new Set(Object.keys(data));
      const descKeys = new Set(Object.keys(dataDescription));
      for (const key of dataKeys) {
        if (!descKeys.has(key)) {
          missingDataDescriptions.push(`${stepName}.${key}`);
        }
      }
    }
  }

  // Check for missing abort strings
  const missingAborts: string[] = [];
  for (const abort of flowAborts) {
    if (!stringAborts.has(abort)) {
      missingAborts.push(abort);
    }
  }

  // Aborts are reported under missing_errors, so they must also count toward validity.
  // Malformed (non-object) step entries are surfaced here too, so a parseable-but-bad
  // strings.json yields a structured validation error rather than a thrown exception.
  const allMissingErrors = [...missingErrors, ...missingAborts, ...malformedSteps];

  const valid =
    missingSteps.length === 0 &&
    allMissingErrors.length === 0 &&
    missingDataDescriptions.length === 0;

  return {
    valid,
    missing_steps: missingSteps,
    orphaned_steps: orphanedSteps,
    missing_errors: allMissingErrors,
    missing_data_descriptions: missingDataDescriptions,
  };
}
